### ISSUE TITLE
Unexpected, non-sequential behavior in bytes concatenation

### ISSUE THREAD

**Issue by Hecate2** [CityOfZion/neo3-boa#444]
With neo3-boa 0.8.0, and neo-mamba at commit 873932c8cb25497b90a39b3e327572746764e699 (a latest merging commit), and the following smart contract `ruler_protocol.py`,
```python
from typing import Any, List, Union

from boa3.builtin import NeoMetadata, metadata, public
from boa3.builtin.contract import Nep17TransferEvent, abort
from boa3.builtin.interop.blockchain import get_contract
from boa3.builtin.interop.contract import NEO, GAS, call_contract
from boa3.builtin.interop.runtime import calling_script_hash, check_witness, get_time
from boa3.builtin.interop.storage import delete, get, put, find
from boa3.builtin.type import UInt160

COLLATERAL_TOKEN = b"NEO"
PARITY_TOKEN = b"GAS"

# TODO: dynamically determine the collateral rate (using oracle?)
COLLATERAL_RATE = 3
COLLATERAL_RATE_STR = b'03'
# TODO: dynamically determine the period of loan
LOAN_PERIOD_DAYS = 30  # the loan expires after this number of days
LOAN_PERIOD_MILLISECONDS = LOAN_PERIOD_DAYS * 86400000  # num of days * milliseconds per day
# (Does the optimizer recognize the calculation above as a const? Does this increase GAS cost?)
MINT_RATIO = 7  # approximately 1 NEO is worth 7 GAS
MINT_RATIO_STR = b'07'
# on_transfer = Nep17TransferEvent  # This contract cannot comply with NEP17 because the names of tokens vary

'''
RC_TOKEN_NAME_PATTERN = f'_RC_{COLLATERAL_RATE}_{MINT_RATIO}_{PARITY_TOKEN}_{EXPIRY_TIMESTAMP}
RR_TOKEN_NAME_PATTERN = f'_RR_{COLLATERAL_RATE}_{MINT_RATIO}_{PARITY_TOKEN}_{EXPIRY_TIMESTAMP}
'''

@public
def main() -> str:
    # put('hello', 'world')  # not allowed in main
    # return "This is a test contract which allows you to borrow GAS by collateralizing NEO. \n" \
    #        "This contract does not give you GAS directly. \n" \
    #        "Instead, you should sell the rcTokens offered by the contract for GAS. \n" \
    #        "Remember to pay back the GAS before the expiry time to retrieve your collateralized NEO."
    return "hello"


@metadata
def manifest_metadata() -> NeoMetadata:
    meta = NeoMetadata()
    meta.author = "github.com/Hecate2"
    meta.description = "Ruler Protocol prototype"
    meta.email = "chenxinhao@ngd.neo.org"
    return meta


@public
def get_loan_period_days() -> int:
    """
    Get the loan period of this contract. The borrower is expected to
    pay the parity token back to the contract in this period of time.
    :return: loan period in days
    """
    return LOAN_PERIOD_DAYS


@public
def onNEP17Payment(from_address: UInt160, amount: int, data: Any) -> bytes:
    """
    The invoker of this method requests a loan from the contract.
    Invoker obtains rcToken.
    :param collateral:
    :param parity:
    :return:
    """
    if from_address is not None:
        assert len(from_address) == 20
    if calling_script_hash == NEO:  # the caller wants to borrow GAS by collateralizing NEO
        return mint_rc_token(from_address, amount*MINT_RATIO)
    elif calling_script_hash == GAS:  # the caller wants to pay back the borrowed GAS
        return b''
    return mint_rc_token(from_address, amount * MINT_RATIO)  # test only
    # abort()


def generate_token_suffix() -> bytes:
    current_time = get_time  # int, milliseconds  # timestamp of the latest block
    expiry_time_bytes = (current_time + LOAN_PERIOD_MILLISECONDS).to_bytes()
    return COLLATERAL_RATE_STR+b'  '+MINT_RATIO_STR+b'  '+PARITY_TOKEN+b'  '+expiry_time_bytes


def mint_rc_token(account: UInt160, mint_amount: int) -> bytes:
    """
    Mints new rc tokens.
    :param account: the address of the account that will receive the rc token
    :type account: UInt160
    :param mint_amount: the amount of rc token to be minted
    :type amount: int
    :raise AssertionError: raised if amount is less than or equal to 0
    """
    assert mint_amount > 0
    token_suffix = generate_token_suffix()
    rc_token_name = b'_RC_' + token_suffix
    rr_token_name = b'_RR_' + token_suffix
    rc_amount = get(rc_token_name + account)
    if rc_amount == b'':
        put(rc_token_name + account, mint_amount)
    else:
        put(rc_token_name + account, mint_amount + rc_amount.to_int())
    rr_amount = get(rr_token_name + account)
    if rr_amount == b'':
        put(rr_token_name + account, mint_amount)
    else:
        put(rr_token_name + account, mint_amount + rr_amount.to_int())
    return rc_token_name


@public
def balanceOf_rc_token(account: UInt160, rc_token_name: bytes) -> int:
    return get(rc_token_name + account).to_int()
@public
def balanceOf_rr_token(account: UInt160, rr_token_name: bytes) -> int:
    return get(rr_token_name + account).to_int()
```
In `def generate_token_suffix`, we concatenate bytes. Compiling and testing it with the following code **(The library `tests` in the following code means neo-mamba/tests, in case there may be an import error. If the latest code of neo-mamba does not work with this test, commenting out the following two lines may help)**:
```python
# comment out the following two lines in neo-mamba/neo3/manifest
        if len(json['features']) != 0:
            raise ValueError("Manifest features is reserved and cannot have any content at this time")
```
```python
import json
from typing import Tuple, List

from neo3 import vm, contracts, storage
from neo3.network import payloads
from neo3.core import to_script_hash, types, cryptography
from neo3.contracts import ApplicationEngine
from tests.contracts.interop.utils import test_engine, test_tx
from copy import deepcopy
from boa3.builtin.interop.contract import NEO, GAS

def read_raw_nef_and_raw_manifest(nef_path: str, manifest_path: str) -> Tuple[bytes, dict]:
    with open(nef_path, 'rb') as f:
        raw_nef = f.read()
    with open(manifest_path, 'r') as f:
        raw_manifest = json.loads(f.read())
    return raw_nef, raw_manifest


def build_nef_and_manifest_from_raw(raw_nef: bytes, raw_manifest: dict) -> Tuple[
    contracts.NEF, contracts.manifest.ContractManifest]:
    nef = contracts.NEF.deserialize_from_bytes(raw_nef)
    manifest = contracts.manifest.ContractManifest.from_json(raw_manifest)
    return nef, manifest


def invoke_main(nef: contracts.NEF) -> ApplicationEngine:
    engine = test_engine(has_container=True, has_snapshot=True)
    engine.load_script(vm.Script(nef.script))
    engine.execute()
    if engine.state == engine.state.FAULT:
        print(f'engine fault from method "main":')
        print(engine.exception_message)
    return engine


def invoke_method(raw_nef: bytes, method: str, params: List[vm.StackItem] = [], signer: str = '',
                  scope: payloads.WitnessScope = payloads.WitnessScope.CALLED_BY_ENTRY,
                  engine:ApplicationEngine=None) -> ApplicationEngine:
    '''
    TODO: multiple signers
    Args:
        signer: ScriptHash of wallet address, "6d629e44cceaf8722c99a41d5fb98cf3472c286a"
    '''
    if not engine:
        engine = test_engine(has_container=True, has_snapshot=True, default_script=False)
    contract = contracts.ContractState(0, nef, manifest, 0, types.UInt160.deserialize_from_bytes(raw_nef))
    try:
        engine.snapshot.contracts.put(contract)
    except ValueError:  # the contract has already been deployed to the blockchain
        pass
    engine.load_script(vm.Script(contract.script))
    if signer:
        engine.script_container.signers = [payloads.Signer(
            types.UInt160.from_string(signer),
            scope
        )]
    array = vm.ArrayStackItem(engine.reference_counter)
    if params:
        array.append(params)  # parameters for method
    engine.push(array)
    engine.push(vm.IntegerStackItem(15))  # CallFlag
    engine.push(vm.ByteStringStackItem(method))
    engine.push(vm.ByteStringStackItem(contract.hash.to_array()))
    engine.invoke_syscall_by_name("System.Contract.Call")
    engine.execute()
    if engine.state == engine.state.FAULT:
        print(f'engine fault from method "{method}":')
        print(engine.exception_message)
    return engine


def print_results(engine:ApplicationEngine):
    print(engine.state, [item.__str__() for item in engine.result_stack._items])


raw_nef, raw_manifest = read_raw_nef_and_raw_manifest('ruler_protocol.nef', 'ruler_protocol.manifest.json')
nef, manifest = build_nef_and_manifest_from_raw(raw_nef, raw_manifest)
contract_owner_hash = "6d629e44cceaf8722c99a41d5fb98cf3472c286a"
contract = contracts.ContractState(0, nef, manifest, 0, types.UInt160.deserialize_from_bytes(raw_nef))
contract_hash = contract.hash

print_results(invoke_main(nef))
deployed_engine = invoke_method(raw_nef, "get_loan_period_days", signer=contract_owner_hash)
print_results(deployed_engine)
# deployed_engine.call_from_native(NEO, contract_hash, "onNEP17Payment", [contract_owner_hash, vm.IntegerStackItem(100), vm.IntegerStackItem(100)])
deployed_engine = invoke_method(raw_nef, 'onNEP17Payment',
                            params=[vm.ByteStringStackItem(types.UInt160.from_string(contract_owner_hash).to_array()),
                                    vm.IntegerStackItem(100), vm.ByteStringStackItem(b'')], signer=NEO, engine=ApplicationEngine(contracts.TriggerType.APPLICATION, payloads.Transaction._serializable_init(),
                             deployed_engine.snapshot, 0, test_mode=True))
print_results(deployed_engine)
rc_token_name = deployed_engine.result_stack._items[0]
deployed_engine = invoke_method(raw_nef, "balanceOf_rc_token",
    params=[vm.ByteStringStackItem(types.UInt160.from_string(contract_owner_hash).to_array()),
                                           rc_token_name.convert_to(vm.StackItemType.BYTESTRING)], engine=ApplicationEngine(contracts.TriggerType.APPLICATION, payloads.Transaction._serializable_init(),
                             deployed_engine.snapshot, 0, test_mode=True))
print_results(deployed_engine)
```
The testing code prints:
```
VMState.HALT ['68656c6c6f']
VMState.HALT ['30', '68656c6c6f']
VMState.HALT ['5f52435f88b29889560147415330333037202020202020', '68656c6c6f']
VMState.HALT ['700', '68656c6c6f']
```
Let's pay special attention to the hex string in the 3rd line of the result, related to the following function excepted from the contract:
```python
def generate_token_suffix() -> bytes:
    current_time = get_time  # int, milliseconds  # timestamp of the latest block
    expiry_time_bytes = (current_time + LOAN_PERIOD_MILLISECONDS).to_bytes()
    return COLLATERAL_RATE_STR+b'  '+MINT_RATIO_STR+b'  '+PARITY_TOKEN+b'  '+expiry_time_bytes
```
The utf-8 string representation of the hex `5f52435f88b29889560147415330333037202020202020` is `_RC_\x88\xb2\x98\x89V\x01GAS0307      `. This can be obtained through `bytes.fromhex('5f52435f88b29889560147415330333037202020202020')`
Now, change the code in `def generate_token_suffix`:
```python
def generate_token_suffix() -> bytes:
    current_time = get_time  # int, milliseconds  # timestamp of the latest block
    expiry_time = current_time + LOAN_PERIOD_MILLISECONDS
    return COLLATERAL_RATE_STR+MINT_RATIO_STR+PARITY_TOKEN+expiry_time.to_bytes()
```
The hex result of the test is `5f52435f3033303747415388b298895601`, which means `_RC_0307GAS\x88\xb2\x98\x89V\x01`.
Another experiment
```python
def generate_token_suffix() -> bytes:
    current_time = get_time  # int, milliseconds  # timestamp of the latest block
    expiry_time_bytes = (current_time + LOAN_PERIOD_MILLISECONDS).to_bytes()
    return COLLATERAL_RATE_STR+b'__'+MINT_RATIO_STR+b'__'+PARITY_TOKEN+b'__'+expiry_time_bytes
```
gives `5f52435f88b298895601474153303330375f5f5f5f5f5f` and `_RC_\x88\xb2\x98\x89V\x01GAS0307______`

I do not really know whether this is more related to neo-vm or neo3-boa. Perhaps this is not even a bug, but just something that is worth a note.

**Comment by Hecate2** [comment]
The sequence concatenated with `+` would be in order if every single term is `bytearray`.

**Comment by meevee98** [comment]
> In `def generate_token_suffix`, we concatenate bytes. Compiling and testing it with the following code **(The library `tests` in the following code means neo-mamba/tests, in case there may be an import error. If the latest code of neo-mamba does not work with this test, commenting out the following two lines may help)**:
> 
> ```python
> # comment out the following two lines in neo-mamba/neo3/manifest
>         if len(json['features']) != 0:
>             raise ValueError("Manifest features is reserved and cannot have any content at this time")
> ```
@Hecate2 about this, the issue is related to neo3-boa. The field `features` wasn't included in the manifest in v0.8, but it's fixed in v0.8.1


About the concatenation, I'm currently working on a fix.
@Hecate2 @hal0x2328 it would be nice if you can test if #457 works for you

**Comment by hal0x2328** [comment]
This fixed the issue for my example
